lib/daru/category.rb

```diff
@@ -0,0 +1,447 @@
+      # Create a hash to map each category to positional indexes
+      categories = data.each_with_index.group_by(&:first)
+      @cat_hash = categories.map { |cat, group| [cat, group.map(&:last)] }.to_h
+
+      # Map each category to a unique integer for effective storage in @array
+      map_cat_int = categories.keys.each_with_index.to_h
+
+      # Inverse mapping of map_cat_int
+      @map_int_cat = map_cat_int.invert
+
+      # To link every instance to its category,
+      # it stores integer for every instance representing its category
+      @array = map_cat_int.values_at(*data)
+    end
+
+    def category_from_position position
+      @map_int_cat[@array[position]]
+    end
+
+    def assert_ordered operation
+      # Change ArgumentError to something more expressive
+      raise ArgumentError, "Can not apply #{operation.to_s} when vector is unordered. "\
+        'To make the categorical data ordered, use #ordered = true'\
+        unless ordered?
+    end
+
+    def dummy_coding full
+      categories = @cat_hash.keys
+      categories.delete(base_category) unless full
+
+      df = categories.map do |category|
+        dummy_code @cat_hash[category]
+      end
+
+      Daru::DataFrame.new df,
+        index: @index,
+        order: create_names(categories)
+    end
+
+    def dummy_code positions
+      code = Array.new(size, 0)
+      positions.each { |pos| code[pos] = 1 }
+      code
+    end
+
+    def simple_coding full
+      categories = @cat_hash.keys
+      categories.delete(base_category) unless full
+
+      df = categories.map do |category|
+        simple_code @cat_hash[category]
+      end
+
+      Daru::DataFrame.new df,
+        index: @index,
+        order: create_names(categories)
+    end
+
+    def simple_code positions
+      n = @cat_hash.keys.size.to_f
+      code = Array.new(size, -1/n)
+      positions.each { |pos| code[pos] = (n-1)/n }
+      code
+    end
+
+    def helmert_coding(*)
+      categories = @cat_hash.keys[0..-2]
+
+      df = categories.each_index.map do |index|
+        helmert_code index
+      end
+
+      Daru::DataFrame.new df,
+        index: @index,
+        order: create_names(categories)
+    end
+
+    def helmert_code index
+      n = (categories.size - index).to_f
+
+      @array.map do |cat_index|
+        if cat_index == index
+          (n-1)/n
+        elsif cat_index > index
+          -1/n
+        else
+          0
+        end
+      end
+    end
+
+    def deviation_coding(*)
+      categories = @cat_hash.keys[0..-2]
+
+      df = categories.each_index.map do |index|
+        deviation_code index
+      end
+
+      Daru::DataFrame.new df,
+        index: @index,
+        order: create_names(categories)
+    end
+
+    def deviation_code index
+      last = categories.size - 1
+      @array.map do |cat_index|
+        case cat_index
+        when index then 1
+        when last  then -1
+        else 0
+        end
+      end
+    end
```

Comment:
This set of pairs XXX_code/XXX_coding methods looks pretty suspicious to me (especially considering the fact that `_coding` counterparts are typically _almost_ the same). Can't it be somehow reduced?..